src/Microsoft.AspNetCore.SystemWebAdapters/HttpApplication.cs

```diff
@@ -0,0 +1,318 @@
+    {
+        add => AddEvent(value);
+        remove => RemoveEvent(value);
+    }
+
+    public event EventHandler? UpdateRequestCache
+    {
+        add => AddEvent(value);
+        remove => RemoveEvent(value);
+    }
+
+    public event EventHandler? PostUpdateRequestCache
+    {
+        add => AddEvent(value);
+        remove => RemoveEvent(value);
+    }
+
+    public event EventHandler? LogRequest
+    {
+        add => AddEvent(value);
+        remove => RemoveEvent(value);
+    }
+
+    public event EventHandler? PostLogRequest
+    {
+        add => AddEvent(value);
+        remove => RemoveEvent(value);
+    }
+
+    public event EventHandler? EndRequest
+    {
+        add => AddEvent(value);
+        remove => RemoveEvent(value);
+    }
+
+    public event EventHandler? Error
+    {
+        add => AddEvent(value);
+        remove => RemoveEvent(value);
+    }
+
+    public event EventHandler? Disposed
+    {
+        add => AddEvent(value);
+        remove => RemoveEvent(value);
+    }
+
+    [Diagnostics.CodeAnalysis.SuppressMessage("Usage", "CA1816:Dispose methods should call SuppressFinalize", Justification = Constants.ApiFromAspNet)]
+    public void Dispose()
+    {
+        InvokeEvent(ApplicationEvent.Disposed);
+
+        foreach (var module in Modules.Modules)
+        {
+            module.Dispose();
+        }
+    }
+
+    internal event EventHandler? SessionStart
+    {
+        add => AddEvent(value);
+        remove => RemoveEvent(value);
+    }
+
+    internal event EventHandler? SessionEnd
+    {
+        add => AddEvent(value);
+        remove => RemoveEvent(value);
+    }
+
+    internal event EventHandler? ApplicationStart
+    {
+        add => AddEvent(value);
+        remove => RemoveEvent(value);
+    }
+
+    internal event EventHandler? ApplicationInit
+    {
+        add => AddEvent(value);
+        remove => RemoveEvent(value);
+    }
+
+    public event EventHandler? PreSendRequestHeaders
+    {
+        add => AddEvent(value);
+        remove => RemoveEvent(value);
+    }
+
+    private void AddEvent(EventHandler? handler, [CallerMemberName] string? name = null)
+    {
+        if (handler is null)
+        {
+            return;
+        }
+
+        if (Enum.TryParse<ApplicationEvent>(name, out var eventName))
+        {
+            AddEvent(eventName, handler);
+        }
+        else
+        {
+            throw new ArgumentOutOfRangeException(nameof(name));
+        }
+    }
+
+    private void RemoveEvent(EventHandler? handler, [CallerMemberName] string? name = null)
+    {
+        if (handler is null)
+        {
+            return;
+        }
+
+        if (Enum.TryParse<ApplicationEvent>(name, out var eventName))
+        {
+            RemoveEvent(eventName, handler);
+        }
+        else
+        {
+            throw new ArgumentOutOfRangeException(nameof(name));
+        }
+    }
+
+    private void AddEvent(ApplicationEvent appEvent, EventHandler handler)
```

Comment:
This doesn't need to be thread safe?

Reply:
Nope - any instance of HttpApplication will be assigned to a request. If a new instance is required, then the initialization will occur on the request it is assigned to.

Reply:
btw HttpApplication is per request, but the HttpApplicationState is shared across all requests

Reply:
I remember now 😬